### ISSUE TITLE
New API: onScrollEndReached (x/y)

### ISSUE THREAD

**Issue by alexkrolick** [gridgrid/grid-react-adapter#5]
A prop that takes a callback for reaching the end of the scrolling area would be useful for infinite scroll/autopagination

**Comment by scamden** [comment]
actually the grid's default data-model can handle scroll virtualized to the server. the suggested model is to load just id's to represent each row or at least to know the row count ahead of time (even with huge data sets this is usually doable) and then fetch the full objects as the grid asks for them. the ReactGrid takes a `loadRows` prop which is `(rowIndexes : number[]) => void` and expects you to populate the data for those rowIndexes if that makes sense. (the indexes are in the data space, that is to say they don't include the header)

**Comment by scamden** [comment]
if you really want to do something on scroll the grid does fire numerous events through it's eventLoop; one is 'grid-cell-scroll'. currently this have to be bound directly through the grid object but i have been considering add event props like react does (`on-grid-cell-scroll={}` etc)

**Comment by alexkrolick** [comment]
Ok, I can look into `loadRows`; however, a prop that fires when the end is reached is more useful in the case where the full number of rows is unknown or only known approximately.

**Comment by scamden** [comment]
Makes sense. Wouldn’t be hard to add either. I will mention that adding or
removing rows is a lot more expensive than changing the data on those rows
because it forces the grid to recalculate a bunch of its internal models.
I’ll go ahead and implement this feature when I get I chance though but
just wanted to give that warning for your info.

**Comment by yashshah12** [comment]
Hey, I am a colleague of @alexkrolick and I was wondering if there is a timeline as to when this feature will be into the next release?

**Comment by scamden** [comment]
well as i told @alexkrolick this feature is not the recommended way to virtualize the data in the grid, so i would strongly recommend taking the approach i mentioned near the top of this issue. if you're dead set on following a different approach i can prioritize this though as it won't take terribly long. let me know.

**Comment by alexkrolick** [comment]
To clarify, is this how it works:

1. Initial data load
2. Create existing row data structure
3. Fill in up to `total` with empty data
4. Get ref to grid (see README)
5. Attach `loadRows` to a function that fetches `(endIndex - startIndex)` more rows and changes the data

What is the page size for the loadRows method? Is it configurable?

**Comment by scamden** [comment]
@alexkrolick that's close ya. there's no need for a grid ref or to fill empty data. you can do it like this

1. initital data load get `total` item count
2. create `total` row objects
3. pass a function to `loadRows` prop that when called fetches the data associated with the supplied row indexes
4. when the data comes back from the fetch either put it on the data field of the correct row object you created, or you can pass all the data you know about directly to the `data` prop of react grid as a two dimensional array where `data[row][col] = gridDataResult`;

The paging is an optimized algorithm that loosely tries to fetch twice the current viewport size (in rows). I'd be open to discussing tweaks to this, but it worked very nicely for us over at salesforceiq, so i'd recommend trying it first to see how it performs for you.

**Comment by alexkrolick** [comment]
Oh, there's a loadRows prop 😁 ... https://github.com/gridgrid/grid-react-adapter#props-and-formats

The query is relatively expensive (it's filtered Elasticsearch data), so I think making fewer would be better.

**Comment by scamden** [comment]
haha :)

ya we also used elastic search. so your filter is on the fetch of the
actual row data? obviously not suggesting you rearchitect your entire stack
haha, but our approach was to run the filter on the ids only query and then
fetch the data directly from those ids. that way you only have to run the
filter query once, instead of on every scroll. just musing really as it's
probably not a feasibly quick change.

def interested to discuss how you'd like to paging to work in an ideal
world. maybe open an issue for that specifically?

**Comment by scamden** [comment]
oh i see how would be issue to miss in the types. the reactgridprops interface extends from the gridopts interface so it has every field there as a prop as well. i'll add it to the docs.

**Comment by alexkrolick** [comment]
>  run the filter on the ids only query
We could do that, it's not a big change for us right now.

**Comment by scamden** [comment]
nice! ya def faster i think, but that's my humble opinion

**Comment by yashshah12** [comment]
Hey @scamden, I see the `loadRows` in the readme and I passed that props to the component, but it didnt call my function. I am just wondering if that function is implemented cause I am looking at your code, and I dont see that `loadRows` prop being used.

**Comment by scamden** [comment]
load rows is used directly by the grid so it's passed along in opts to the
grid create method. it definitely should be calling you. can you post a
code snippet?

**Comment by alexkrolick** [comment]
It's not working because the extra rows aren't populated with ids. We'll take this offline and check back.

**Comment by scamden** [comment]
ok i had a feeling that might be it :) let me know :)
